File: 2019/LevelUpDecBatch/l007_HashMapAndHeap/basicHM.java
import java.util.HashMap;
import java.util.ArrayList;

public class basicHM {
    public static void test1() {
        HashMap<String, Integer> map = new HashMap<>();
        map.put("USA", 1000);
        map.put("IND", 10000);
        map.put("NEP", 90);
        map.put("usa", 9000);

        map.put("USA", 9990);
        System.out.println(map.get("Usa"));
        System.out.println(map);
        for (String s : map.keySet())
            System.out.println(s + " -> " + map.get(s));
    }

    // Frequency map of given string
    public static void test2(String str) {
        HashMap<Character, Integer> map = new HashMap<>();
        // for (int i = 0; i < str.length(); i++) {
        // char ch = str.charAt(i);
        // if (!map.containsKey(ch))
        // map.put(ch, 1);
        // else
        // map.put(ch, map.get(ch) + 1);
        // }

        for (int i = 0; i < str.length(); i++) {
            char ch = str.charAt(i);
            map.put(ch, map.getOrDefault(ch, 0) + 1);
        }
    }

    public static void test3(String str) {
        HashMap<Character, ArrayList<Integer>> map = new HashMap<>();
        for (int i = 0; i < str.length(); i++) {
            char ch = str.charAt(i);
            map.putIfAbsent(ch, new ArrayList<>());
            map.get(ch).add(i);
        }

        System.out.println(map);
    }

    public static void main(String[] args) {
        test3("abcaksjbabaabadabab");

    }

}

File: 2019/LevelUpDecBatch/l007_HashMapAndHeap/basicHM.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <unordered_map>

using namespace std;

void test1()
{
    unordered_map<string, int> map;
    map["USA"] = 1000;
    map["IND"] = 10000;
    map["NEP"] = 90;
    map["usa"] = 9000;

    // map["Usa"] = 9000;
    cout << map["USA"] << endl;
    if (map.find("usa") != map.end())
        cout << map["usa"] << endl;

    for (pair<string, int> p : map)
        cout << p.first << " -> " << p.second << endl;
}

// Frequency map of given string
void test2(string str)
{
    unordered_map<char, int> map;
    for (char ch : str)
    {
        map[ch]++;
    }
}

// Index of all character
void test3(string str)
{
    unordered_map<char, vector<int>> map;
    for (int i = 0; i < str.length(); i++)
    {
        char ch = str[i];
        map[ch].push_back(i);
    }

    for (auto &key : map)
    {
        char k = key.first;
        vector<int> val = key.second;
        cout << k << " -> ";
        for (int ele : val)
            cout << ele << ", ";
        cout << endl;
    }
}

int main()
{
    test3("abcaksjbabaabadabab");
    return 0;
}
